### ISSUE TITLE
lint: namedConst false positives and false negatives

### ISSUE THREAD

**Issue by Quasilyte** [go-critic/go-critic#330]
This issue summarizes known `namedConst` check false positives and false negatives.

**False positives**

1. Const comparison with it's underlying value. Used for consts defined with iota to catch whether new values were added to that list.
```go
// BAD: should not suggest to replace 7 with PDISCARD
if PDISCARD != 7 {
	panic("PDISCARD changed; does all Class values still fit in three bits?")
}
```
2. `time.Weekday` used in arith expressions:
```go
// BAD: replacing 3 with time.Wednesday is not correct
day := days[3*t.Weekday():]
```
> Note: `time.Duration` handled in adhoc way. We may extend that exception list or find a general solution to this.

3. OS perms.

```
$GOPATH/src/upspin.io/cmd/upspinfs/upspinfs_test.go:100:44: namedConst: use os.ModePerm instead of 511
```

```
$GOPATH/src/github.com/zenazn/goji/web/regexp_pattern.go:119:44: namedConst: use syntax.POSIX instead of 0
```

4. Circular definitions.

```go
// Protocol versions.
const (
	PvInvalid ProtocolVersion = 0
	PvV3      ProtocolVersion = 3
	PvV4      ProtocolVersion = 4
	PvMax     ProtocolVersion = PvV4
)
```
It suggests to replace 4 with `PvMax`, but it would lead to const decl loop.

**False negatives**

1. Sometimes time expressions may benefit from named const suggestions:
```go
// d is time.Duration
// Could suggest time.Second instead of 1e9
dsec := int64(d / 1e9)

60*60*1e9 // time.Hour?
```
